### PR TITLE
Allow `CoreMeter` implementations to use their own custom attribute reference

### PR DESCRIPTION
**Pull request by Sushisource**
## What was changed
Added `CustomMetricAttributes` trait / variant to `MetricAttributes` to allow langs to have a simple reference based approach to allocating their own attributes.

## Why?
Allows destroying backing attributes on `Drop` of the thing implementing `CustomMetricAttributes`

## Checklist
It may no longer be worth keeping a buffer at all, or, if we do, I need to re-implement it to work in terms of refs.

1. Closes 

2. How was this tested:
Existing tests

3. Any docs updates needed?

### CODE REVIEW

**Review by cretz — APPROVED** [review]

**Review by mjameswh — APPROVED** [review]

File: core-api/src/telemetry/metrics.rs [hunk 1]
```diff
@@ -197,14 +164,92 @@ pub trait Gauge: Send + Sync {
...
+}
+
+/// A lazy reference to some metrics buffer attributes
+pub type BufferAttributes = LazyRef<Arc<dyn CustomMetricAttributes + 'static>>;
+
+/// Types lang uses to contain references to its lang-side defined instrument references must
+/// implement this marker trait
+pub trait BufferInstrumentRef {}
+/// A lazy reference to a metrics buffer instrument
+pub type LazyBufferInstrument<T> = LazyRef<Arc<T>>;
```
Comment (cretz) [inline core-api/src/telemetry/metrics.rs:214]: Can we put a type constraint on `T` that it has to implement `BufferInstrumentRef`? Or maybe `I` can have that constraint above? Or rather, is it at least expected that `T` implements `BufferInstrumentRef`?

File: core/src/telemetry/metrics.rs [hunk 2]
```diff
@@ -637,44 +628,66 @@ pub fn start_prometheus_metric_exporter(
...
         }
     }
 }
-impl CoreMeter for MetricsCallBuffer {
-    fn new_attributes(&self, opts: MetricsAttributesOptions) -> MetricAttributes {
-        let id = self.attribute_ids.fetch_add(1, Ordering::AcqRel);
+
+impl<I> CoreMeter for MetricsCallBuffer<I>
+where
+    I: Debug + Send + Sync + Clone + 'static,
```
Comment (cretz) [inline core/src/telemetry/metrics.rs]: Hrmm, so I guess if I'm tying my non-cloneable Python metric's instrument instance to this, I will have to wrap it in Arc or similar? It doesn't seem like `CustomMetricAttributes` has to implement `Clone`. I don't think it's a problem though.

**Review by cretz — APPROVED** [review]

File: core-api/src/telemetry/metrics.rs (hunk 1 above)
Comment (cretz) [reply core-api/src/telemetry/metrics.rs:214]: EDIT: Nevermind, I see this constraint is now available on the `CoreMeter` impl

File: core/src/telemetry/metrics.rs (hunk 2 above)
Comment (Sushisource) [reply core/src/telemetry/metrics.rs]: The trait doesn't require it, but the places where it's used do (when they need it) which is fairly typical practice

**Pull request merged**